Watch did stop on Test Cluster

The watch for workspace events has stopped on the test cluster. There was no exception logged. 

Our manual timer based watch logged some exceptions (see below), so it might have something to do with the connection to the cluster API. 

We should try to add more logging to figure out how to detect this state, so that we may set up the watches again. 

```
io.fabric8.kubernetes.client.KubernetesClientException: Operation: [list]  for kind: [TemplateSpecResource]  with name: [null]  in namespace: [theiacloud]  failed.
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:130) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:138) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:413) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:402) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:83) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.theia.cloud.operator.KillAfterRunnable.run(KillAfterRunnable.java:60) [operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.net.ConnectException: Failed to connect to /10.52.0.1:443
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:249) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:258) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.execute(RealCall.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl.send(OkHttpClientImpl.java:138) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.retryWithExponentialBackoff(OperationSupport.java:575) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:554) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:519) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:503) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:131) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	... 10 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:?]
	at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source) ~[?:?]
	at java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source) ~[?:?]
	at java.net.AbstractPlainSocketImpl.connect(Unknown Source) ~[?:?]
	at java.net.SocksSocketImpl.connect(Unknown Source) ~[?:?]
	at java.net.Socket.connect(Unknown Source) ~[?:?]
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:129) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:247) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:258) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.execute(RealCall.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl.send(OkHttpClientImpl.java:138) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.retryWithExponentialBackoff(OperationSupport.java:575) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:554) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:519) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:503) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:131) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	... 10 more
```

--- a/java/common/org.eclipse.theia.cloud.common/.classpath
+++ b/java/common/org.eclipse.theia.cloud.common/.classpath
@@ -6,20 +6,9 @@
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry excluding="**" kind="src" output="target/classes" path="src/main/resources">
-		<attributes>
-			<attribute name="maven.pomderived" value="true"/>
-		</attributes>
-	</classpathentry>
-	<classpathentry kind="src" output="target/test-classes" path="src/test/java">
-		<attributes>
-			<attribute name="optional" value="true"/>
-			<attribute name="maven.pomderived" value="true"/>
-			<attribute name="test" value="true"/>
-		</attributes>
-	</classpathentry>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11">
 		<attributes>
+			<attribute name="module" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
@@ -30,9 +19,9 @@
 	</classpathentry>
 	<classpathentry kind="src" path="target/generated-sources/annotations">
 		<attributes>
+			<attribute name="ignore_optional_problems" value="true"/>
 			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
-			<attribute name="ignore_optional_problems" value="true"/>
 			<attribute name="m2e-apt" value="true"/>
 		</attributes>
 	</classpathentry>
@@ -43,17 +32,17 @@
 	</classpathentry>
 	<classpathentry kind="src" output="target/test-classes" path=".apt_generated_tests">
 		<attributes>
-			<attribute name="optional" value="true"/>
 			<attribute name="test" value="true"/>
+			<attribute name="optional" value="true"/>
 		</attributes>
 	</classpathentry>
 	<classpathentry kind="src" output="target/test-classes" path="target/generated-test-sources/test-annotations">
 		<attributes>
+			<attribute name="ignore_optional_problems" value="true"/>
+			<attribute name="test" value="true"/>
 			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
-			<attribute name="ignore_optional_problems" value="true"/>
 			<attribute name="m2e-apt" value="true"/>
-			<attribute name="test" value="true"/>
 		</attributes>
 	</classpathentry>
 	<classpathentry kind="output" path="target/classes"/>

--- a/java/operator/org.eclipse.theia.cloud.operator/log4j2.xml
+++ b/java/operator/org.eclipse.theia.cloud.operator/log4j2.xml
@@ -11,6 +11,10 @@
 			additivity="false">
 			<AppenderRef ref="Console" />
 		</Logger>
+		<Logger name="org.eclipse.theia.cloud.operator.KillAfterRunnable" level="warn"
+			additivity="false">
+			<AppenderRef ref="Console" />
+		</Logger>
 		<Logger name="org.eclipse.theia.cloud.operator" level="trace"
 			additivity="false">
 			<AppenderRef ref="Console" />

--- a/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/SpecWatch.java
+++ b/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/SpecWatch.java
@@ -90,4 +90,16 @@ final class SpecWatch<S extends CustomResource<?, ?>> implements Watcher<S> {
 		cause);
 	System.exit(-1);
     }
+
+    @Override
+    public void onClose() {
+	LOGGER.info(formatLogMessage(correlationIdPrefix, resourceName + " watch closed"));
+	Watcher.super.onClose();
+    }
+
+    @Override
+    public boolean reconnecting() {
+	LOGGER.info(formatLogMessage(correlationIdPrefix, resourceName + " reconnecting"));
+	return Watcher.super.reconnecting();
+    }
 }

